### PR TITLE
refactor(api): slightly faster simulation

### LINKED ISSUE #9503: bug: Opentrons app lists wrong protocol steps
# Overview

The in-app simulation lists wrong steps, even impossible ones, which makes it harder to follow the protocol while running. This issue happened before 5.0.0. as well. This doesn’t happen when simulating with opentrons_simulate on the terminal.

# Steps to reproduce
- Use 364-well plates and a multi-channel pipette
- **Use custom 200 uL tips instead of the default 300 uL ones**
- Include a loop such as the following in the Python protocol

```
for r in range(3): # 1 reservoir well --> 6 columns of the plate ; 3 wells total                                            
    d = r*6
    p300multi.distribute(100,    
                         reservoir.wells()[r+first_r_well].bottom(mm), 
                         plate.columns()[0+d : 6+d],                   
                         blow_out=False, 
                         disposal_volume=0,    
                         new_tip="never")
```

# Current behavior

The app states that the pipette is aspirating 300 uL (which is impossible with the 200 uL tips), distributing 100+100+100 and aspirating again. However, what it really is doing is aspirate 200 uL and fill one of the 364 columns before aspirating again (distributes 100 uL in the A row + distributes 100 uL in the B row).

image

Before running, the whole list of steps is wrong. When it starts running, once the step has been performed it gets "corrected" like this

image

As you can see, it's still wrong.

# Expected behavior

The app should list the correct steps

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

TLDR using caching to reduce simulation duration. These changes apply both to fast simulation and slow simulation.

While digging into #9503 I considered that maybe we're better off just improving the speed of slow simulation and doing away with fast simulation all together. Why? Because it seemed agonizing to have the `PipetteDict` produced by fast simulation match what is produced by the HC. 

From profiling a Protocol that has thousands of aspirate/dispenses I learned that most of the SynchAdapter price was paid in the `move_to` function. It's called repeatedly. So I attempted to modify our `InstrumentContextSimulation` to have it inherit from `InstrumentContextImplementation` and ONLY override `move_to`. So basically, use the backward strategy from how fast sim had worked.

I spent a good deal of time profiling simulation and found sneaky little functions that are called zillions of times for no good reason. Comparing results on the PI I got:

```
  Slow Sim - 0:01:42.165820 
  New Fast sim - 0:00:56.065534
  Original Fast Sim - 0:00:33.754647
```

Pretty good, but not good enough. 

So I fixed #9503 and am creating this PR to add the profiler learnings to improve simulation times.

# Changelog

- lru cache mathy functions. These are slow and called repeatedly.
- Add `thermocycler_present` property to `Deck`. `should_dodge_thermocycler` is called zillions of times and does a linear search for thermocycler in deck map. We can cache this value.
- BAD_PAIRS can be a set to avoid linear search

# Review requests

Am I caching stuff that I shouldn't? Are the lru_cache limits reasonable?

*I will add real life performance improvement metrics when I can*. 

# Risk assessment

The `thermocycler_present` attribute is somewhat risky, though it is well covered by unit tests.
Otherewise. Low.

### CODE REVIEW

File: api/src/opentrons/hardware_control/instrument_handler.py [hunk 1]
```diff
@@ -393,19 +396,22 @@ def ready_for_tip_action(self, target: Pipette, action: HardwareAction) -> None:
...
     ) -> float:
         mm = ul / instr.ul_per_mm(ul, action)
         position = mm + instr.config.bottom
         return round(position, 6)
 
+    @functools.lru_cache(PLUNGER_CACHE_LIMIT)
     def plunger_speed(
         self, instr: Pipette, ul_per_s: float, action: "UlPerMmAction"
     ) -> float:
         mm_per_s = ul_per_s / instr.ul_per_mm(instr.config.max_volume, action)
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/instrument_handler.py]: `ul_per_mm` calls `piecewise_volume_conversion` which does a decent bit of math.

These `plunger_X` methods can be called thousands of times with very small variance in inputs. The config values are immutable so caching seems like a save move.

File: api/src/opentrons/protocols/geometry/planning.py [hunk 2]
```diff
@@ -61,7 +60,7 @@ def should_dodge_thermocycler(
 
     Returns True if we need to dodge, False otherwise
     """
-    if any([isinstance(item, ThermocyclerGeometry) for item in deck.values()]):
```
Comment (amitlissack) [inline api/src/opentrons/protocols/geometry/planning.py:64]: This method is called in each `plan_moves` which is called in each `move_to`. A needless search for a value that can be cached.

File: api/src/opentrons/protocols/geometry/deck.py [hunk 3]
```diff
@@ -67,6 +69,7 @@ def _load_fixtures(self):
             self.__setitem__(slot_name, loaded_f)
 
     @staticmethod
+    @functools.lru_cache(20)
```
Comment (sfoster1) [inline api/src/opentrons/protocols/geometry/deck.py:72]: I wonder if it might be better to apply this to `_check_name`, which appears to be the only caller of `_assure_int` and does at least some more work - in the worst case, it'll do both an `_assure_int` call and a `__contains__` call which does a set op and so on and so on. I guess we'd need to invalidate the cache whenever `self.data` changes though.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
`InstrumentHandler` has the same lifetime as the hardware control api object and takes the attached instruments as inputs. That means that we need to clear the caches on the plunger position/speed/flowrate when the attached instruments change. 

What might be a better idea is adding an LRU cache on `hardware_control.pipettes.Pipette.ul_per_mm`. Since that has the lifetime of an instrument, and it gets replaced when you change instruments, it won't need cache clearing, and it's where all the work happens.

I also think we could change the cache location for the deck slot key checking, but reasonable people can disagree.

File: api/src/opentrons/protocols/geometry/deck.py (hunk 3 above)
Comment (amitlissack) [reply api/src/opentrons/protocols/geometry/deck.py:72]: > `InstrumentHandler` has the same lifetime as the hardware control api object and takes the attached instruments as inputs. That means that we need to clear the caches on the plunger position/speed/flowrate when the attached instruments change.
> 
> What might be a better idea is adding an LRU cache on `hardware_control.pipettes.Pipette.ul_per_mm`. Since that has the lifetime of an instrument, and it gets replaced when you change instruments, it won't need cache clearing, and it's where all the work happens.
> 
> I also think we could change the cache location for the deck slot key checking, but reasonable people can disagree.

Thanks

**Comment by amitlissack** [comment]
> `InstrumentHandler` has the same lifetime as the hardware control api object and takes the attached instruments as inputs. That means that we need to clear the caches on the plunger position/speed/flowrate when the attached instruments change.
> 
> What might be a better idea is adding an LRU cache on `hardware_control.pipettes.Pipette.ul_per_mm`. Since that has the lifetime of an instrument, and it gets replaced when you change instruments, it won't need cache clearing, and it's where all the work happens.
> 
> I also think we could change the cache location for the deck slot key checking, but reasonable people can disagree.

I'm not sure I understand you regarding `InstrumentHandler`. If the attached instruments change won't the `instr` arg to the plunger methods be different hence creating a new cache entry?

I was hoping to do exactly what you suggested, but lru cache requires hashable arguments. `ul_per_mm` and `_check_names` fail that test. I will continue working on this.

File: api/src/opentrons/protocols/geometry/deck.py [hunk 4]
```diff
@@ -75,6 +77,7 @@ def _assure_int(key: object) -> int:
         else:
             raise TypeError(type(key))
 
+    @functools.lru_cache(20)
```
Comment (mcous) [inline api/src/opentrons/protocols/geometry/deck.py]: Is this safe? It looks like `self.data` will mutate without busting the cache

File: api/src/opentrons/protocols/geometry/deck.py [hunk 5]
```diff
@@ -295,3 +305,12 @@ def get_item_covered_slot_keys(sk, i):
...
         return colliding_items
+
+    @property
+    def thermocycler_present(self) -> bool:
+        """Is a thermocycler present on the deck."""
+        return self._thermocycler_present
+
+    def __hash__(self) -> int:
+        """A hash function as UserDict is unhashable."""
+        return id(self)
```
Comment (mcous) [inline api/src/opentrons/protocols/geometry/deck.py]: This feels iffy for the same reason the cache above feels iffy

**Comment by amitlissack** [comment]
> > `InstrumentHandler` has the same lifetime as the hardware control api object and takes the attached instruments as inputs. That means that we need to clear the caches on the plunger position/speed/flowrate when the attached instruments change.
> > What might be a better idea is adding an LRU cache on `hardware_control.pipettes.Pipette.ul_per_mm`. Since that has the lifetime of an instrument, and it gets replaced when you change instruments, it won't need cache clearing, and it's where all the work happens.
> > I also think we could change the cache location for the deck slot key checking, but reasonable people can disagree.
> 
> I'm not sure I understand you regarding `InstrumentHandler`. If the attached instruments change won't the `instr` arg to the plunger methods be different hence creating a new cache entry?
> 
> I was hoping to do exactly what you suggested, but lru cache requires hashable arguments. `ul_per_mm` and `_check_names` fail that test. I will continue working on this.

I was wrong. It's `piecewise_volume_conversion` whose args cannot be hashed. Moving cache to `ul_per_mm`

**Review by mcous — COMMENTED** [review]
Caching on a `Deck` method while the state of `Deck` can mutate underneath it makes me a little anxious. Is it necessary given the addition of the `thermocycler_preset` property?

**Comment by amitlissack** [comment]
> Caching on a `Deck` method while the state of `Deck` can mutate underneath it makes me a little anxious. Is it necessary given the addition of the `thermocycler_preset` property?

I had initially put the lru_cache on `_assure_int` which is static, potentially called thousands of times, and harmless to cache. I moved the cache to `_check_names` without thinking it through.

**Review by mcous — APPROVED** [review]

**Review by sfoster1 — APPROVED** [review]
Looks great to me!
